### ISSUE TITLE
Javascript type definitions (d.ts) incorrect for unlabelled constructor arguments

### ISSUE THREAD

**Issue by JohnBjrk** [gleam-lang/gleam#1998]
The type definitions generated for constructors with unlabelled arguments are not valid typescript.

For example the definition om  `Some` in stdlib becomes the following:
```gleam
export class Some<FH> extends _.CustomType {
  constructor(0: FH);
  //----------^
  
  x0: FH;
}
```

Where `0` is not a valid name for an identifier.

A suggested solution (I did a quick experiment) is to add a prefix `p` to the unlabelled arguments 
which would yield the following type definition.

```gleam
export class Some<FH> extends _.CustomType {
  constructor(p0: FH);
  //----------^
  
  x0: FH;
}
```

I don't think this should affect anything else since this is just the name in the type definition and I don't think it is actually referenced outside the actual definition.

**Comment by lpil** [comment]
Nice find! We'd need to use a prefix with a `$` in it to avoid colliding with a user specified `p1` field. `argument$1` or such.

**Comment by JohnBjrk** [comment]
I was trying to use my patch locally in my test ts project and I think there is one more thing that needs to be fixed. The type definition exposes a field which for the `Some` type is called x0, but the underlying js implementation actually stores the data as `this[0]`, which means that when using the type definition for implementing a ts ffi, instance of the type will seem to have a `instanceOfSome.x0`, but this is undefined.

My initial thoughts is that this could probably be fixed by naming the field `0` in the type-definition. I'll try it out some more and update the original PR #1999 if I find a working solution.

**Comment by JohnBjrk** [comment]
I think this issue can be closed once #1999 has been merged, but I'm thinking about opening another one about adding some integration test where we actually use the type-definitions. Not sure if its feasible though - I guess that would need a project that depends on the typescript compiler and has some minimal typescript file which uses the types produced by gleam.

**Comment by lpil** [comment]
Doing that on CI sounds great! Thanks
